zebra-state/src/service.rs

```diff
@@ -711,27 +682,102 @@ impl Service<Request> for StateService {
+            Request::AwaitUtxo(outpoint) => {
+                metrics::counter!(
+                    "state.requests",
+                    1,
+                    "service" => "state",
+                    "type" => "await_utxo",
+                );
+
+                let timer = CodeTimer::start();
+
+                // Prepare the AwaitUtxo future.
+                let response_fut = self.pending_utxos.queue(outpoint);
+                let span = Span::current();
+                let response_fut = response_fut.instrument(span).boxed();
+
+                // Check the non-finalized block queue.
+                if let Some(utxo) = self.queued_blocks.utxo(&outpoint) {
+                    self.pending_utxos.respond(&outpoint, utxo);
+
+                    // We're finished, the returned future gets the UTXO from the respond() channel.
+                    timer.finish(module_path!(), line!(), "AwaitUtxo/queued-non-finalized");
+
+                    return response_fut;
+                }
+
+                // We ignore any UTXOs in FinalizedState.queued_by_prev_hash,
+                // because it is only used during checkpoint verification.
+                //
+                // This creates a rare race condition, but it doesn't seem to happen much in practice.
+                // See #5126 for details.
+
+                // Send a request to the ReadStateService, to get UTXOs from the finalized chain.
+                let read_service = self.read_service.clone();
+
+                // Optional TODO:
+                //  - make pending_utxos.respond() async using a channel,
+                //    so we can use ReadRequest::ChainUtxo here, and avoid a block_in_place().
+
+                // Check any non-finalized chain, and the finalized chain.
+                let span = Span::current();
+                let utxo = tokio::task::block_in_place(move || {
+                    span.in_scope(move || {
+                        read_service.non_finalized_state_receiver.with_watch_data(
+                            |non_finalized_state| {
+                                read::any_utxo(non_finalized_state, &read_service.db, outpoint)
+                            },
+                        )
+                    })
+                });
+
+                if let Some(utxo) = utxo {
+                    self.pending_utxos.respond(&outpoint, utxo);
+
+                    // We're finished, the returned future gets the UTXO from the respond() channel.
+                    timer.finish(module_path!(), line!(), "AwaitUtxo/any-chain");
+
+                    return response_fut;
+                }
+
+                // We're finished, but the returned future is waiting on the respond() channel.
+                timer.finish(module_path!(), line!(), "AwaitUtxo/waiting");
+
+                response_fut
+            }
+
+            // TODO: add a name() method to Request, and combine all the generic read requests
+            //
```

Comment:
Following the suggestion to add the metrics::counter at the top of `.call()`

```suggestion
            Request::Depth(_)
            | Request::Tip
            | Request::BlockLocator
            | Request::Transaction(_)
            | Request::Block(_)
            | Request::FindBlockHashes { .. }
            | Request::FindBlockHeaders { .. } => {
                // Redirect the request to the concurrent ReadStateService
                let read_service = self.read_service.clone();
                async move {
                    let req = req
                        .try_into()
                        .expect("ReadRequest conversion should not fail");
                    let rsp = read_service.oneshot(req).await?;
                    let rsp = rsp.try_into().expect("Response conversion should not fail");
                    Ok(rsp)
                }
                .boxed()
            }

```

Reply:
These changes can also go in that separate PR.